File: src/main/java/com/github/tetrisanalyzer/piecemove/PieceMove.java
package com.github.tetrisanalyzer.piecemove;

import com.github.tetrisanalyzer.board.Board;
import com.github.tetrisanalyzer.move.Move;
import com.github.tetrisanalyzer.piece.Piece;
import com.github.tetrisanalyzer.piece.Point;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

/**
 * Has responsible of setting and clearing a piece on the board and can tell which
 * moves are valid on the (none empty) board.
 */
public class PieceMove {
    private int pieceHeight;
    private Board board;
    private Piece piece;
    private Move move;

    private PieceMove down;
    private Set<PieceMove> asideAndRotate = new HashSet<PieceMove>();

    private int[] boardLineIndices;
    private int[] orLines;
    private int[] andLines;

    private static final int ALL_BITS_CLEARED = 0;
    private static final int ALL_BITS_SET = -1;

    public PieceMove(Board board, Piece piece, Move move) {
        this.board = board;
        this.piece = piece;
        this.move = move;

        pieceHeight = piece.height(move.getRotation());
        orLines = new int[pieceHeight];
        andLines = new int[pieceHeight];
        Arrays.fill(orLines, ALL_BITS_CLEARED);
        Arrays.fill(andLines, ALL_BITS_SET);
        boardLineIndices = new int[pieceHeight];

        for (int y=0; y<pieceHeight; y++) {
            boardLineIndices[y] = move.getY() + y;
        }
        for (Point point: piece.getShape(move.getRotation()).getPoints()) {
            orLines[point.getY()] |= (1 << (move.getX() + point.getX()));
        }
        for (Point point: piece.getShape(move.getRotation()).getPoints()) {
            andLines[point.getY()] &= ~(1 << (move.getX() + point.getX()));
        }
    }

    public Move getMove() {
        return move;
    }

    public PieceMove getDown() {
        return down;
    }

    public void setDown(PieceMove down) {
        this.down = down;
    }

    public void addPieceMove(PieceMove pieceMove) {
        asideAndRotate.add(pieceMove);
    }

    public Set<PieceMove> getAsideAndRotate() {
        return asideAndRotate;
    }

    /**
     * Sets a piece on the board.
     *
     * @return number of cleared lines
     */
    public int setPiece() {
        for (int y=0; y<pieceHeight; y++) {
            board.setBits(boardLineIndices[y], orLines[y]);
        }
        return board.clearLines(move.getY(), pieceHeight);
    }

    /**
     * Removes a piece from the board.
     */
    public void clearPiece() {
        for (int y =0; y<pieceHeight; y++) {
            board.clearBits(boardLineIndices[y], andLines[y]);
        }
    }

    /**
     * True if the piece position is not occupied.
     */
    public boolean isFree() {
        for (int y=0; y<pieceHeight; y++) {
            if (!board.isBitsFree(boardLineIndices[y], orLines[y])) {
                return false;
            }
        }
        return true;
    }

    /**
     * True if this piece hasn't reached the bottom (down == null) and the move down
     * (current position where y is increased by one) is not occupied with "dots" (rest of pieces).
     */
    public boolean canMoveDown() {
        return down != null && down.isFree();
    }

    public Set<PieceMove> getFreeAsideAndRotateMoves() {
        Set<PieceMove> freeMoves = new HashSet<PieceMove>();

        for (PieceMove pieceMove : asideAndRotate) {
            if (pieceMove.isFree()) {
                freeMoves.add(pieceMove);
            }
        }
        return freeMoves;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        PieceMove pieceMove = (PieceMove) o;

        if (move != null ? !move.equals(pieceMove.move) : pieceMove.move != null) return false;
        if (piece != null ? !piece.equals(pieceMove.piece) : pieceMove.piece != null) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = piece != null ? piece.hashCode() : 0;
        result = 31 * result + (move != null ? move.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return piece + ", " + move;
    }
}

File: src/main/scala/nu/tengstrand/tetrisanalyzer/piecemove/PieceMove.scala
package nu.tengstrand.tetrisanalyzer.piecemove

import nu.tengstrand.tetrisanalyzer.board.Board
import nu.tengstrand.tetrisanalyzer.piece.Piece
import nu.tengstrand.tetrisanalyzer.move.Move
import collection.immutable.{Set, HashSet}

object PieceMove {
  val AllBitsCleared = 0

  def apply(board: Board, piece: Piece, move: Move): PieceMove = {
    val AllBitsSet = -1

    val pieceHeight = piece.height(move.rotation)
    val orRows = Array.fill(pieceHeight) { AllBitsCleared }
    val andRows = Array.fill(pieceHeight) { AllBitsSet }
    val boardRowIndices = Array.ofDim[Int](pieceHeight)

    for (y <- 0 until pieceHeight)
      boardRowIndices(y) = move.y + y

    piece.shape(move.rotation).dots.foreach(point => orRows(point.y) |= (1 << (move.x + point.x)))
    piece.shape(move.rotation).dots.foreach(point => andRows(point.y) &= ~(1 << (move.x + point.x)))

    new PieceMove(board, piece, move, boardRowIndices, orRows, andRows)
  }
}

/**
 * Has responsible of setting and clearing a piece on the board and can tell which
 * moves are valid on the (none empty) board.
 */
class PieceMove(val board: Board, val piece: Piece, val move: Move, boardRowIndices: Array[Int],
                orRows: Array[Int], andRows: Array[Int]) extends Ordered[PieceMove] {
  var down: PieceMove = null
  var asideAndRotate: Set[PieceMove] = new HashSet[PieceMove]
  private val pieceHeight = piece.height(move.rotation)

  var animatedPath: PieceMove = null
  private var animatedPathValue: Int = Integer.MAX_VALUE

  /**
   * Sets a piece on the board, return number of cleared rows.
   */
  def setPiece(): Int = {
    for (y <- 0 until pieceHeight) {
      board.rows(boardRowIndices(y)) |= orRows(y)
    }
    board.clearRows(move.y, pieceHeight)
  }

  /**
   * Removes a piece from the board.
   */
  def clearPiece() {
    for (y <- 0 until pieceHeight) {
      board.rows(boardRowIndices(y)) &= andRows(y)
    }
  }

  /**
   * True if the piece position is not occupied.
   */
  def isFree: Boolean = {
    var y = 0
    while (y < pieceHeight && ((board.rows(boardRowIndices(y)) & orRows(y)) == PieceMove.AllBitsCleared)) {
      y += 1
    }
    y == pieceHeight
  }

  /**
   * True if this piece hasn't reached the bottom (down == null) and the move down
   * (current position where y is increased by one) is not occupied with "dots" (pieces parts).
   */
  def canMoveDown: Boolean = {
    down != null && down.isFree
  }

  def freeAsideAndRotateMoves = {
    asideAndRotate.filter(_.isFree)
  }

  /**
   * Resets the path values so the animated path can be calculated by calling calculateAnimatedPath.
   */
  def prepareAnimatedPath() {
    if (animatedPathValue != Integer.MAX_VALUE) {
      animatedPathValue = Integer.MAX_VALUE
      asideAndRotate.foreach(_.prepareAnimatedPath())
      if (down != null)
        down.prepareAnimatedPath()
    }
  }

  /**
   * Calculates an animated path which goes backwards, linked via the reference
   * calculateAnimatedPath, from the move "at the bottom" up to the starting position.
   *
   * Make sure the method prepareAnimatedPath has been called before calling this method.
   */
  def calculateAnimatedPath(fromPieceMove: PieceMove, pathValue: Int, asideValue: Int) {
    if (pathValue + asideValue < animatedPathValue && isFree) {
      animatedPath = fromPieceMove
      animatedPathValue = pathValue + asideValue
      asideAndRotate.foreach(pieceMove => pieceMove.calculateAnimatedPath(this, animatedPathValue, asideValue + 1))
      if (canMoveDown) {
        down.calculateAnimatedPath(this, animatedPathValue, 0)
      }
    }
  }

  def compare(that: PieceMove) = {
    if (move.rotation.compare(that.move.rotation) != 0)
      move.rotation.compare(that.move.rotation)
    else if (move.x.compare(that.move.x) != 0)
      move.x.compare(that.move.x)
    else
      move.y.compare(that.move.y)
  }

  override def equals(that: Any) = that match {
    case other: PieceMove => piece == other.piece && move == other.move
    case _ => false
  }

  override def hashCode = piece.hashCode * 31 + move.hashCode

  override def toString = piece + ", " + move
}